lib/puppet/settings/ini_file.rb

```diff
@@ -169,6 +169,10 @@ def write(fh)
       fh.write(value)
       fh.puts(suffix)
     end
+
+    def ==(other)
+      super(other) && self.line_number == other.line_number
+    end
```

Comment:
Not directly related to the issue at hand, but should we apply the same logic to the `SectionLine` below? As in, does puppet get confused if there are two sections with the same name?

Reply:
I was thinking about that briefly while making the change, but I don't think we really need to worry about that.

When looking to update a setting we either grab the existing `SettingLine` for that key, or try to find the "last" `SettingLine` in the appropriate section. Finding the "last" `SettingLine` in the appropriate section will end up grabbing the last line from the last instance of a section. The only time we end up trying to directly add after a `SectionLine` would be if all instances of the section were empty. In which case, it shouldn't matter which one we end up adding the new `SettingLine` to (though we do add it to the first one).

In a more general sense, I'm not sure that it matters that we be able to differentiate one `SectionLine` from another? We don't have any actions that operate on a section itself. The closest I can think of would be limiting the scope of `SettingLine` operations (in which case, we always want the union of the `SettingLine` entries in the `SectionLine`s, and this already is what happens). For example, I don't think it makes sense to try to do something like "delete the 3rd occurrence of the '`agent`' section, and all settings in it". The closest I could think of would be "remove all occurrences of the '`master`' section (and all settings in them)".

So...any reasonable operation I can think of that involves a `SettingLine` is an all-or-nothing (Eg: scoping, removal of section), or doesn't care which one we're talking about (Eg: Updating an existing setting, adding a new one). That said, if you are thinking of a case where we might care about knowing which instance of a section we're talking about (beyond line number in the conf file, which is handled completely independently already), I'm definitely curious to hear about it.